### ISSUE TITLE
upgrades: (cortex|dd|loki)-apis pods crash looping

### ISSUE THREAD

**Issue by sreis** [opstrace/opstrace#450]
**Describe the bug**

(cortex|dd|loki|opstrace)-apis pods crash looping with the following error

```
time="2021-03-05T19:22:41Z" level=error msg="API_AUTHTOKEN_VERIFICATION_PUBKEY_SET must not be empty. Exit."
```

`loki-api` spec of a crash looping pod:

```
spec:
  progressDeadlineSeconds: 600
  replicas: 2
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: loki-api
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        k8s-app: loki-api
    spec:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchLabels:
                  k8s-app: loki-api
              topologyKey: kubernetes.io/hostname
            weight: 100
      containers:
      - args:
        - -listen=:8080
        - -tenantname=system
        - -loki-querier-url=http://querier.loki.svc.cluster.local:1080
        - -loki-distributor-url=http://distributor.loki.svc.cluster.local:1080
        env:
        - name: API_AUTHTOKEN_VERIFICATION_PUBKEY_SET
        image: opstrace/loki-api:b3c0c6cb64ad8d38e65e2a576a4f84c3e2d8981b
        imagePullPolicy: Always
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /metrics
            port: 8080
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        name: loki-api
        ports:
        - containerPort: 8080
          name: http
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /metrics
            port: 8080
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: loki-api
      serviceAccountName: loki-api
      terminationGracePeriodSeconds: 30
```

**To Reproduce**

Upgrade from `opstrace/controller:d7fc806b-dev` to `opstrace/controller:1a8998dfd-dev`.

**Comment by jgehrcke** [comment]
Thanks for creating the ticket! Great. OK. While I didn't think of this specific failure path, this error is actually perfectly expected and understood.

I would love to relate this to two important discussions / talking points.

1. This upgrade is of the scenario "from any commit to any commit". I was checking back with us if supporting this is our goal, and we very much tended to saying that it is not. We should ask ourselves again if this maybe is our goal, and we should ask ourselves what this topic means with respect to merging to `main`! We might find that this is still not our goal, but also that we -- as of now -- can't reliably test for upgrade breakage before merging to main. And that would mean that **we should always assume that an upgrade from wherever to current main _is_ broken** (we probably assume that already, but it's important to state explicitly :)).
2. **What's not tested is broken** has been confirmed here, great! Was revealing this specific kind of breakage in the `loadcluster` environment an efficient approach? Did we we really expect things to fall apart @sreis or did you expect this to go smooth? Just checking. Maybe we should explore if the `loadcluster` env is the right env to discover "obvious" breakages like that. (layered approach to upgrade testing? from small to bigger complexity?)

Maybe a pragmatic approach here is the following:
* We should always assume that an upgrade from wherever to current main _is_ broken. 
* We may then, ad-hoc, make it our goal that an upgrade from a specific commit to another specific commit (the target commit) _should_ work. We can then set out and _make it work_, assuming that this requires extra work, i.e. assuming it currently _does not work_. That is, the target commit will after all _not_ be the actual target commit. It's _that_, plus _only_ those changes that are required to make the upgrade work. An obvious method to do this w/o slowing down merges to `main` is to put the required patches into a branch.

I think we can and should take the more heavy-weight "releases" topic out of the upgrade discussion again for now. We will simply set the goal  "now, let's make a build, that we can upgrade to" (for the current high-prio SM target environment), and then do the necessary testing and patching work (in a branch).

**Comment by sreis** [comment]
> What's not tested is broken has been confirmed here, great! Was revealing this specific kind of breakage in the loadcluster environment an efficient approach? Did we we really expect things to fall apart @sreis or did you expect this to go smooth? Just checking. Maybe we should explore if the loadcluster env is the right env to discover "obvious" breakages like that. (layered approach to upgrade testing? from small to bigger complexity?)

No, I did not think things would go smooth. After #441, we already knew we needed to do some manual steps. The purpose of the test run last Friday was to understand what, if anything, we needed to patch. The outcome was this issue and #451. 

I'll move the discussion about the upgrade procedure to separate tickets.

**Comment by jgehrcke** [comment]
Thanks @sreis!

**Comment by jgehrcke** [comment]
I believe this specific upgrade challenge has been fixed with #452. Closing. Let's reopen if my thinking turns out to be naively wrong :-).
